### ISSUE TITLE
missing/useful [snake~] methods

### ISSUE THREAD

**Issue by porres** [pure-data/pure-data#1996]
an issue to discuss and propose more [snake~] methods

- [snake~ nchs], get number of channels in a multichannel connection, useful to route signals according to number of channels and also to resize [clone], see https://github.com/pure-data/pure-data/issues/1989
- [snake~ mix], this method could have a 2nd inlet with a toggle and one could use it to mix all channels into one when on, or pass it unchanged otherwise. This would be an alternative to https://github.com/pure-data/pure-data/issues/1994
- [snake~ ltos] "list to sig" or something. Create a multichannel according to a list input. A bit like [sig~] but for lists and multichannels, it would also allow us to create a variable number of channels length.

**Comment by umlaeute** [comment]
what's the advantage of `[snake~ ltos]` over the enhanced version of `[sig~]` as suggested by @Spacechild1 ?

**Comment by porres** [comment]
> what's the advantage of [snake~ ltos] over the enhanced version of [sig~] as suggested by @Spacechild1 ?

being able to create multichannel signals with a different number of channels according to the list input length

**Comment by Spacechild1** [comment]
I think these are all useful, but I would prefer different names:

`[snake~ nchs] -> `[snake~ count]` or `[snake channels]`
`[snake~ mix]` -> `[snake~ sum]`
`[snake~ ltos]` -> `[snake~ fromlist]`

Two more methods:

* `[snake~ split]`: split multichannel signal into two (the split index can be set dynamically)

* `[snake~ merge]` or `[snake~ join]`: merg/join two multi-channel signals together

**Comment by abreubacelar** [comment]
since split feels like the verb from list split, merge could be [snake~ append] and [snake~ prepend] ?

**Comment by Spacechild1** [comment]
> merge could be [snake~ append] and [snake~ prepend] ?

Prepend/append only makes sense for control objects where you have hot and cold inlets. For signal objects you only need a single method.

**Comment by abreubacelar** [comment]
oh, understand, i thought about avoiding crossed connections, but it is silly xD

and what about an object to take specific channels from a signal?

not sure about the verb... get/take/reorder... ?

[snake~ get 1 3 2 4] could take an arbitrary signal with n channels and swap 3 and 2 channels and return just a 4 channel signal from the n channel input, would this be possible with the current objects?

maybe this could even be used for a specific duplication of signals 

[snake~ get 1 2 1 3] could take a 3 channel, and output a 4 channel with the third signal being a copy from the first signal from the input

if the number of channels is lesser than the number expected, so just insert an empty signal there

and we could also drop some channels

like [snake~ get 4 2 6 3] ignoring 1 and 5 and reordering everything

what do you think?

**Comment by Spacechild1** [comment]
> and what about an object to take specific channels from a signal?

+1

**Comment by whale-av** [comment]
I miss "Magic Glass" more than anything else from Pd Extended.  An enhanced version capable of showing multichannel values (and control values) would be useful.
If that is not possible, then maybe an enhanced [print~]...?
Also (sorry) in the Pd Help List of Objects.... patch, separating the audio objects that are multichannel capable from those that are not would be helpful.
David.

**Comment by umlaeute** [comment]
@whale-av while all of your suggestions are related to multichannel, none of them is related to `[snake~] itself...so they seem to be a bit ot for this issue. you probably want to create a separate ticket.

**Comment by whale-av** [comment]
I apologize... I clicked the wrong thread link after registering to post.
Thank you all for all your great work.

**Comment by porres** [comment]
interleave and deinterleave methods would also be cool, just made externals for these in ELSE

**Comment by umlaeute** [comment]
what would they do? how would they be useful (in the context of Pd, where we deal with non-interleaved signals).

**Comment by porres** [comment]
> what would they do? 

I stole the idea from MAX's mc.interleave~ and mc.deinterleave~ hope this is clear:

<img width="296" alt="Image" src="https://github.com/user-attachments/assets/54b84f22-b3af-4778-bed0-1f429750b6ea" />
<img width="250" alt="Image" src="https://github.com/user-attachments/assets/dd47d431-7cbf-4e28-b4d3-b226fa9801de" />

seems useful to me so you can group many, say, stereo files and join then will all the left channels and right channels grouped. Of couse you can also thinkg of quadraphonic/octaphonic and whatnot.

Though, honestly, I never used any of this, like not even other MC tools I am now providing in ELSE for a while now (which are basically all the functions here requestes: join/split/whatever).

**Comment by ben-wes** [comment]
would it make sense to split additional requirements for [snake~] into multiple issues? feels like some of these still require more discussion than others and possibly, there will also be several accompanying PRs then? and i assume that they'd be easier to discuss without losing the context then.

i currently see:

1. get number of channels (already has #1989)
   `[snake~ count]` (or `nchs`, `nchans`, `channels`, `size`)
2. sum mc signal into single channel signal
   `[snake~ sum]` (or `mix`)
3. create mc signal of dynamic size from list
   `[snake~ fromlist]` (or `ltos`)
4. pick specific channels from mc signal
   `[snake~ get]` (or `pick`, `take`, `reorder`)
5. split mc signal at given index
   `[snake~ split]`
6. merge two mc signals into one
   `[snake~ merge]` (or `join`)
7. fill up (or clip) mc signal to specific channel count (referring to https://github.com/pure-data/pure-data/pull/2058#issuecomment-2907727252 - not sure if still needed?)
   `[snake~ fill]` or (or `clip`, `wrap`? possibly better verbs can be found)
8. (de)interleave mc signals
   `[snake~ interleave]` / `[snake~ deinterleave]` (or `lace` / `delace`)

**Comment by umlaeute** [comment]
> * merge two mc signals into one
>   `[snake~ merge]` (or `join`)

how about `[snake~ in]` also allows multichannel intputs?

```
[sig~ 1]    [sig~ 2] [sig~ 3] [sig~ 4]
|           |        |        |
[snake~ in 2]        [snake~ in 2]
|                    |
[snake~     in     2 ]
|
[snake~ out 4]
|
```

> 8. (de)interleave mc signals
>         `[snake~ interleave]` / `[snake~ deinterleave]` (or `lace` / `delace`)

as a non-native speaker is still find these verbs confusing. (if a native speaker thinks they are OK, then i'll probably have to adapt...)

**Comment by Spacechild1** [comment]
> how about [snake~ in] also allows multichannel intputs?

I actually thought about this. I still prefer a dedicated method because it makes the intent more clear, but I would also be fine with the `[snake~ in]` solution.

**Comment by ben-wes** [comment]
> how about `[snake~ in]` also allows multichannel inputs?

sounds good to me! first thought was that it might be an unexpected feature for this learnt method. but seems obvious on second thought.

> 5\. split mc signal at given index

was also wondering if this could optionally be multiple indices (resulting in n+1 outlets)?

> > 8\. `[snake~ interleave]` / `[snake~ deinterleave]` (or `lace` / `delace`)
>
> as a non-native speaker i still find these verbs confusing [...]

zip / unzip? (inspired by https://book.pythontips.com/en/latest/zip.html)

**Comment by ben-wes** [comment]
besides the PR mentioned above (which contains a simple test patch now in the description), i attempted to add `split` and `pick` as well on https://github.com/ben-wes/pure-data/tree/feature/snake-split-pick

... some questions in this context:
* i added a `channels` message to `pick`. it doesn't seem wrong, but it could also be a second inlet expecting a list. any thoughts on this? and possibly, an empty list should reset to pass-through again (not the case yet)?
* similarly, the `split` method currently has a `split` message. not sure about that name (feels wrong to repeat the verb somehow). and maybe it should really be a second inlet like for `list split` in this case?
* `split` currently expects a 0-based split index. makes sense to me - but at the same time, channels are counted from 1 ... so not completely sure here.

**Comment by ben-wes** [comment]
@porres i'm responding to https://github.com/pure-data/pure-data/pull/2615#issuecomment-2961094084 here. i'd prefer to keep discussion on functionality here and use the PR discussion mainly for implementation details.

> [else/sum~] object works like this, you can change its state and tell it not to sum as well. It's useful for me as I found the need to sometimes decide wether I want to sum or not.

sounds useful to me - although i admittedly never had this use case so far. do you have an example?

i'm now also wondering whether summing might be a new message/functionality for [snake~ in]? feels a bit weird to me if a `sum` method can be used for not summing. hmm ... on the other hand, passing the signal through is currently also an option for [snake~ pick] in the branch mentioned above. curious about other opinions here. maybe, something like `bypass 1/0` might be a more generic message that can be used in other occasions as well?

**Comment by Spacechild1** [comment]
> i'm now also wondering whether summing might be a new message/functionality for [snake~ in]?

Please no. `[snake~ in]` has a specific job: create a multichannel signal from multiple input signals. Let's not overload it with unrelated functionality.

**Comment by ben-wes** [comment]
i included this bypass feature for snake_sum~ now in the branch above. it's just a suggestion - but feels ok to me:
* `-b` flag activates it on creation 
* similarly, a `bypass 1/0` message can be used to switch it on/off

**Comment by ben-wes** [comment]
current state for further discussion:

![Image](https://github.com/user-attachments/assets/8e625bba-9750-4de7-bc5d-adbdb1452b45)

**Comment by porres** [comment]
> i included this bypass feature for snake_sum~ now in the branch above. it's just a suggestion - but feels ok to me

looks good

> i'm now also wondering whether summing might be a new message/functionality for [snake~ in]? 

I know this has been discarded, but I'll stress, the point is to deal with a s singal when you already get it as a MC one... ;) and a sum method to [snake~ in] would be as useful as using no object at all and just connecting many outputs together in a single inlet :) 

> do you have an example?

not on the top of my head, but I've been definitely using it in [else/sum~] and this feels like a useful feature overall.

--------

"zip"/"unzip" looks weird to me. Interleave/deinterleave are pretty standard concepts, but I see there's a reference in programming languages and they're much better to type in :)

-------

just to get things straight, now [snake~ in] can also take and join MC signals? Can you provide an example?

**Comment by porres** [comment]
About count, it only works with bang input, right? So one would need to include [r pd-dsp-started] to work with it, huh?

I just feel it's natural that the object would automatically spit out the number of channels when it changes, that's how I did with [else/nchs~]

**Comment by ben-wes** [comment]
> and a sum method to [snake~ in] would be as useful as using no object at all

that's true for the "old" [snake~ in] which is not intended for mc input. but with the updated one, you could use [snake~ in 1] for manipulating mc input if this message was available there. anyway - i discarded that and agree with @Spacechild1 that it's better to have different verbs for different concerns.

> "zip"/"unzip" looks weird to me

i don't have strong preferences concerning these verbs - but yeah, it's concise and not wrong. :)

> now [snake~ in] can also take and join MC signals? Can you provide an example?

yes. with this, it gains the functionality of a `join` method. see https://github.com/pure-data/pure-data/issues/1996#issuecomment-2957028729 or the 2 channel input on the left inlet in my screenshot above. i attached this patch in the description of the PR.

> About count, it only works with bang input, right? So one would need to include [r pd-dsp-started] to work with it, huh?

true. see https://github.com/pure-data/pure-data/pull/2615#discussion_r2136627271 ... to me, seeing this in line with `samplerate~` seems logical.

**Comment by umlaeute** [comment]
i just noticed that the mc capabilities of `[snake~ in]` seem to not mix well with non-multichannel signals:

<img width="447" height="140" alt="Image" src="https://github.com/user-attachments/assets/f4ecbca6-29ff-44e2-8602-ac22c6b78dfd" />

i would have expected the left-hand output to be `[1 2 3 4\` rather than `[1 2 3 0\`

**Comment by ben-wes** [comment]
> i would have expected the left-hand output to be `[1 2 3 4\` rather than `[1 2 3 0\`

for sure, yes. just tested and got different results. looking into it ...

**Comment by ben-wes** [comment]
> i would have expected the left-hand output to be `[1 2 3 4\` rather than `[1 2 3 0\`

should be fixed now in #2615.

**Comment by fdch** [comment]
I've taken a stab at a weighed sum object for the snake family here https://github.com/fdch/pure-data/tree/snake_wsum-proposal. I've included also the help patch modification. I think it relates to the "mix" type of family suggested here, but it is indeed a weighted sum. 

The name [snake~ wsum] is a placeholder, but it is indeed a "weighted sum", but it could use a better name.

One thing I'm worried here is that there are no other "perform" routines in the snake~ family (except for a print), so maybe that's by design, making this proposal rather moot. In any case, I wanted to ask about the approach here.

If there's interest, I'll make it a proper PR against upstream :) Please let me know what you think!

**Comment by ben-wes** [comment]
> weighted sum

thanks for the proposal and demo! just curious: afaict, a weighted sum can be built quite simply with existing objects since https://github.com/pure-data/pure-data/pull/2615 (and with equal weights by dividing the sum by `[siginfo~ channels]`) - so i'm not sure about the advantage of this rather specific object?

**Comment by fdch** [comment]
thanks for the quick look

well, i have been using the multichannel tabread4~ as a multichannel tabosc4~ (here https://github.com/fdch/fd_lib/blob/main/abstractions/tabosc4m~.pd) and it does come in handy to be able to assign different weights to the sum (besides 1/N) without unpacking and packing every channel

**Comment by ben-wes** [comment]
> unpacking and packing every channel

this should be avoidable with [sig~]:

<img width="238" height="229" alt="Image" src="https://github.com/user-attachments/assets/73bb0d20-d8a5-48d9-90f2-e29c7f5e4dce" />

**Comment by ben-wes** [comment]
since you need variable channel count, this probably doesn't solve your problem, sorry. not sure what the most elegant way is for creating a static signal of dynamic channel count, i admit...

**Comment by fdch** [comment]
I do like your approach, though, but yes, it is not as flexible as an internal solution like the one I suggest. Maybe i need to think a bit more about this. 

The real trigger here was a multichannel tabosc4~, but I could really solve that with an abstraction. I could have naturally cloned tabosc4~, giving each table its own weight before going to the outlet. That's when I realized, it would be great for snake sum to accept a list of weights, making it possible for many other uses as well, hence the proposed object. As a plus, internally these weights are normalized, and the object accepts incomplete lists and attempts to give an intuitive solution for those cases.

**Comment by ben-wes** [comment]
i'm curious about other opinions - to me, this use case seems too specific for a dedicated verb.

[snake~ fromlist] would allow to handle it quite well. since that's not available though, you might have to use an abstraction to have that functionality. i'm attaching my approach (using the new [list foreach]) here: [list2sig.zip](https://github.com/user-attachments/files/32481286/list2sig.zip)

but yeah: then you could also handle everything within the clone instance.

~... and to me, multichannel support for tabosc4\~ seems like a valid request, too.~

**Comment by umlaeute** [comment]
> ... and to me, multichannel support for tabosc4~ seems like a valid request, too.

well, aren't we back to https://github.com/pure-data/pure-data/issues/2947 then?

**Comment by ben-wes** [comment]
> well, aren't we back to [#2947](https://github.com/pure-data/pure-data/issues/2947) then?

yep, i take that back. when writing it, i thought: why does tabread4~ support mc and not tabosc4~? but there's obviously a reason ...
